### ISSUE TITLE
Directive handling

### ISSUE THREAD

**Issue by porcellus** [rollup/rollup#1855]
Both module level and function level directives are ignored/removed.
[Reproduction](https://rollupjs.org/repl?version=0.53.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwYSgpJTIwJTdCJTVDbiU1Q3QlNUMlMjJ1c2UlMjBhc20lNUMlMjIlNUNuJTVDdCU1Q24lNUN0cmV0dXJuJTIwMSUzQiU1Q24lN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyYW1kJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=)

- On the function level these should be retained: for example it can cause a serious performance degradation in the case of "use asm".
- On the module level I think they should produce an error as simply including them can cause errors or ineffective directives depending on if the directive is included in the first module or not. I see no obvious way of handling these.

If this is ok, I will submit a PR with both new test cases and a fix. I have it ready [here](https://github.com/tresorit/rollup). Just raising this issue for discussion as per the PR message.

**Comment by lukastaegert** [comment]
Nice! Admittedly, I wasn't aware directives were represented in the AST by acorn.
One thing to note: "use strict" is one directive we probably always want to remove as it is implicitly part of the ES6 spec and is added automatically when generating non-ES6 bundles. Otherwise, there might be a lot of superfluous "use strict" in your code that you pull in via dependencies.

As for module level directives: I think they should rather be a warning than an error as in many cases, the code will probably still work so we might be unnecessarily breaking code.

**Comment by porcellus** [comment]
I though so, I already remove them in my code.
I will change the error into a warning, update the tests accordingly and submit my PR

**Comment by alippai** [comment]
Supporting directives is a new feature introduced in October: https://github.com/acornjs/acorn/blob/master/CHANGELOG.md#520-2017-10-30

**Comment by kzc** [comment]
> One thing to note: "use strict" is one directive we probably always want to remove as it is implicitly part of the ES6 spec 

Non-strict mode did not go away in ES6. There's all sorts of crazy code out there in the wild.
```
$ node -e 'a = 1; console.log(a);'
1
```
```
$ node -e '"use strict"; a = 1; console.log(a);'
[eval]:1
"use strict"; a = 1; console.log(a);
                ^
ReferenceError: a is not defined
```
```
$ echo 'var n = 1; n.p = 2; console.log(n);' | node
1
```
```
$ echo '"use strict"; var n = 1; n.p = 2; console.log(n);' | node
[stdin]:1
"use strict"; var n = 1; n.p = 2; console.log(n);
                             ^
TypeError: Cannot create property 'p' on number '1'
```

**Comment by lukastaegert** [comment]
Yes, but it has gone away in ES6 modules and we add we add "use strict" for non-ES6 bundles:
https://stackoverflow.com/questions/31685262/not-recommended-to-use-use-strict-in-es6

**Comment by lukastaegert** [comment]
Closing this as I believe this has now been resolved with the merge of #1856.
